keps/sig-cloud-provider/1959-service-lb-class-field/README.md

```diff
@@ -0,0 +1,213 @@
+within my own network because I want to save costs.
+
+#### Story 2
+
+As an application developer:
+* I MUST use a hardware-based loadbalancer for certain applications due to specific protocols
+only available there.
+* I want to use the cloud provider's default load balancer for any applications that do not
+rely on protocols from hardware load balancers.
+
+### Risks and Mitigations
+
+Many cloud providers today support an "opt-out" annotation for this behavior. The annotation is specific
+to the cloud provider. Introduction of the `loadBalancerClassName` field at this point would mean that
+cloud providers need to start accounting for both existing annotations and the new field.
+
+## Design Details
+
+Introduce a new field to Service `spec.loadBalancerClassName`.
+
+If the field `spec.loadBalancerClassName` is not set, the existing cloud provider will assume
+ownership of the Service Type=LoadBalancer resource. This is required to not break existing clusters
+that assume Service Type=LoadBalancer is always managed by the cloud provider.
+
+Required API changes:
+```go
+// ServiceSpec describes the attributes that a user creates on a service.
+type ServiceSpec struct {
+	...
+	...
+
+	// loadBalancerClassName is the name of the load balancer implementation this Service belongs to.
+	// This field can only be set when the Service type is 'LoadBalancer'. If not set, the default load
+	// balancer implementation is used, today this is typically done through the cloud provider integration,
+	// but should apply for any default implementation. If set, it is assumed that a load balancer
+	// implementation is watching for Services with a matching class name. Any default load balancer
+	// implementation (e.g. cloud providers) should ignore Services that set this field.
+	// +optional
+	LoadBalancerClassName string `json:"loadBalancerClassName,omitempty"`
+}
+```
+
+* `loadBalancerClassName` will be immutable so that existing and future implementations do not have to worry
+about handling Services that change the class name.
+* `loadBalancerClassName` can carry arbitrary string values.
+* the `loadBalancerClassName` field will be feature gated. The field will be dropped during API strategy unless
+the feature gate is enabled.
+
+Required updates to service controller:
+* if the class field is NOT set for a Service, allow the cloud provider to reconcile the load balancer.
+* if the class annotation IS set for a Service, skip reconciliation of the Service from the cloud provider.
+
+### Test Plan
+
+Unit tests:
+* test that service controller does not call the cloud provider if the class field is set.
+* test API strategy to ensure the `loadBalancerClassName` field is dropped unless the feature gate is enabled
+or an existing Service has the field set.
+* test API validation for immutability.
+
+E2E tests:
+* test that creating a Service with an unknown class name results in no load balancer being created for a Service.
```

Comment:
We'll need integration tests to make sure you can change type and the field gets cleared/validated as proper

Reply:
added a section for integration tests